### DISCUSSION TITLE
use cd for directory hyperlinks

### DISCUSSION THREAD

**Discussion by Luticus** [kovidgoyal/kitty#5377] [Q&A] [answered]
Currently the behavior of hyperlink directories, from ls output for instance, is to open my default file manager. I'd much rather it, instead, cd to the clicked directory in my working terminal. Is there a way to do this? Currently I've set up the following in my ~/.config/kitty/open-actions.conf:
protocol file
mime inode/directory
action .....

The issue is that I don't know what to do for the action there. I've tried a few different launch actions, send-text, etc. Nothing will do what I'm hoping to accomplish so far. They all pretty much seem to just do nothing.

Was hoping someone in the community may have already worked this one out. Thanks in advance!

**Comment by kovidgoyal** [comment, marked as answer]
```conf
# Change directory in shell
protocol file
mime inode/directory
action send_text normal,application cd ${FILE_PATH}\r
```

**Reply by Luticus** [reply]
That's amazing!  Thanks!

**Reply by Luticus** [reply]
For anyone looking at this I altered my commands so they would all behave more "in line" and so they look like this:
```
# open text in vi in the same terminal
protocol file
mime text/*
action send_text normal,application vi "${FILE_PATH}"\r
```

```
# show clicked picture inline as if you did an icat in the terminal
protocol file
mime image/*
action send_text normal,application kitty +kitten icat "${FILE_PATH}"\r
```

```
# change to clicked directory and automatically ls
protocol file
mime inode/directory
action send_text normal,application cd "${FILE_PATH}" && ls\r
```

The one *major* downside to these is that they break down when ssh is involved. Basically kitty tries to "open" the directory like it's a text file or something, probably because it's keying off of "protocol file", not sure it there's a possible work around yet.

**Comment by kovidgoyal** [comment]
Use the ssh kitten, then you can just do

edit-in-kitty file

and it will automatically and transparently open the remote file in a
local editor instance, with all the bells and whistles.
https://sw.kovidgoyal.net/kitty/shell-integration/#edit-files-in-new-kitty-windows-even-over-ssh

**Reply by Luticus** [reply]
That's not what I meant. So like if is ssh in and then do ls --hyperlink=auto and then I click on a directory on the remote system it tries to "open" it as if it were a file, which it understandably fails to do. Instead I'd expect it to run the cd "${FILE_PATH} command as if it were local.

**Comment by page-down** [comment]
> ... they break down when ssh is involved ...
> ... if is ssh in and then do ls --hyperlink=auto ...

I noticed that `kitty/window.py` `open_url()` has `remote_file` kitten hardcoded in it, when the file URL contains the hostname (local host name).
So the following will not work.

```text
protocol file
url ^file:\/\/debug-host\/
action ...
```

It would be nice if there is a better way to be able to freely customize the open action. For example `remote_file` kitten should be defined in the default open action, when hostname is not the local host name. This way the user can override the default rule.

> ... click on a directory on the remote system ...

Until this is improved, you can replace the protocol in hyperlink with a custom one and define matching rules.
`mime inode/directory` will probably not work on URLs from remote host.
You need to check if the path is a directory on the remote system and execute cd, otherwise other commands.

**Comment by kovidgoyal** [comment]
On Sat, Aug 13, 2022 at 07:15:59PM -0700, page-down wrote:
> It would be nice if there was a better way to be able to freely customize the open action. For example `remote_file` kitten should be defined in the default open action, when hostname is not empty. This way the user can override the default rule.

Doing that would break backward compatibility, since it would mean the
remote file kitten would not work for anyone that has defined their own
open actions. 

The best that can be done in a backwards compat way is to define a
separate set of open actions that apply for remote URLs, but this
wouldn't solve OP's problem anyway, see below.

> 
> > ... click on a directory on the remote system ...
> 
> Once you can customize the rule that hostname is not empty, you need to check if the path is a directory on the remote system and execute cd, otherwise other commands.

There is no way to know a URL points to a file or a directory on a
remote system, so you cant really use open actions to match against
directories anyway. You would need to change the OSC 8 spec to demand
that URLs to directories end with / and then change ls and other
programs that generate these to comply with that.

**Reply by Luticus** [reply]
I see what you're saying here. Basically because it's a remote system and you cant check if it's a directory it's probably matching one of the other mime types first and treating it as though it should be opened in vim, etc. Am I understanding correctly?

**Comment by kovidgoyal** [comment]
No currently, clicking on links in remote systems is hardcoded to use
the remote files kitten. I am saying that even if that were changed,
without changing ls to output a trailing / on links to directories your
use case is not supportable.

**Reply by Luticus** [reply]
Oh OK. Well it's unfortunate, but understandable  :)

**Reply by mazunki** [reply]
just stumbled upon this. POSIX ls supports `-p`, which appends `/` to directory files. wouldn't that suffice for this?
